File: api/go-http/files/pkg/cache/cache.go
package cache

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"
	"sync"
	"time"

	"github.com/patrickmn/go-cache"
	"github.com/redis/go-redis/v9"
)

// Cache provides caching functionality with Redis (optional) or in-memory fallback.
// Install Redis support: go get github.com/redis/go-redis/v9
//
// Usage:
//     import "{{PROJECT_SLUG}}/apps/{{PLATFORM_SLUG}}/pkg/cache"
//
//     // Set a value (TTL in seconds)
//     cache.Set("user:123", userData, 3600)
//
//     // Get a value
//     user, found := cache.Get("user:123")
//
//     // Delete a value
//     cache.Delete("user:123")
//
//     // Delete all keys matching pattern (supports * wildcard)
//     cache.DeletePattern("user:*")        // All user keys
//     cache.DeletePattern("product:*")     // All product keys

var (
	defaultCache Cache
	once         sync.Once
)

// Cache interface for cache operations
type Cache interface {
	Get(ctx context.Context, key string) (interface{}, bool)
	Set(ctx context.Context, key string, value interface{}, ttl time.Duration)
	Delete(ctx context.Context, key string)
	DeletePattern(ctx context.Context, pattern string)
	Exists(ctx context.Context, key string) bool
	Clear(ctx context.Context)
	Close() error
}

// InMemoryCache is a simple in-memory cache for development/testing
type InMemoryCache struct {
	cache *cache.Cache
	mu    sync.RWMutex
}

// NewInMemoryCache creates a new in-memory cache instance
func NewInMemoryCache() *InMemoryCache {
	return &InMemoryCache{
		cache: cache.New(5*time.Minute, 10*time.Minute),
	}
}

func (c *InMemoryCache) Get(ctx context.Context, key string) (interface{}, bool) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.cache.Get(key)
}

func (c *InMemoryCache) Set(ctx context.Context, key string, value interface{}, ttl time.Duration) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.cache.Set(key, value, ttl)
}

func (c *InMemoryCache) Delete(ctx context.Context, key string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.cache.Delete(key)
}

func (c *InMemoryCache) DeletePattern(ctx context.Context, pattern string) {
	c.mu.Lock()
	defer c.mu.Unlock()

	// Simple wildcard support for in-memory cache
	if strings.Contains(pattern, "*") {
		parts := strings.Split(pattern, "*")
		prefix := parts[0]
		suffix := ""
		if len(parts) > 1 {
			suffix = parts[1]
		}

		// Get all items from the cache
		items := c.cache.Items()

		// Delete matching keys
		for key := range items {
			if strings.HasPrefix(key, prefix) && (suffix == "" || strings.HasSuffix(key, suffix)) {
				c.cache.Delete(key)
			}
		}
	} else {
		// Exact match
		c.cache.Delete(pattern)
	}
}

func (c *InMemoryCache) Exists(ctx context.Context, key string) bool {
	c.mu.RLock()
	defer c.mu.RUnlock()
	_, found := c.cache.Get(key)
	return found
}

func (c *InMemoryCache) Clear(ctx context.Context) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.cache.Flush()
}

func (c *InMemoryCache) Close() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.cache.Flush()
	return nil
}

// RedisCache is a Redis-backed cache for production
type RedisCache struct {
	client *redis.Client
	logger *slog.Logger
}

// NewRedisCache creates a new Redis cache instance
func NewRedisCache(url string, logger *slog.Logger) (*RedisCache, error) {
	opts, err := redis.ParseURL(url)
	if err != nil {
		return nil, fmt.Errorf("invalid redis url: %w", err)
	}

	client := redis.NewClient(opts)

	// Test connection
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := client.Ping(ctx).Err(); err != nil {
		return nil, fmt.Errorf("redis connection failed: %w", err)
	}

	return &RedisCache{
		client: client,
		logger: logger,
	}, nil
}

func (c *RedisCache) Get(ctx context.Context, key string) (interface{}, bool) {
	val, err := c.client.Get(ctx, key).Result()
	if err == redis.Nil {
		return nil, false
	}
	if err != nil {
		c.logger.Error("redis get error", "error", err, "key", key)
		return nil, false
	}

	// Try to unmarshal as JSON
	var result interface{}
	if err := json.Unmarshal([]byte(val), &result); err != nil {
		// If not JSON, return as string
		return val, true
	}
	return result, true
}

func (c *RedisCache) Set(ctx context.Context, key string, value interface{}, ttl time.Duration) {
	var val string
	switch v := value.(type) {
	case string:
		val = v
	default:
		bytes, err := json.Marshal(value)
		if err != nil {
			c.logger.Error("cache set marshal error", "error", err, "key", key)
			return
		}
		val = string(bytes)
	}

	if err := c.client.Set(ctx, key, val, ttl).Err(); err != nil {
		c.logger.Error("redis set error", "error", err, "key", key)
	}
}

func (c *RedisCache) Delete(ctx context.Context, key string) {
	if err := c.client.Del(ctx, key).Err(); err != nil {
		c.logger.Error("redis delete error", "error", err, "key", key)
	}
}

func (c *RedisCache) DeletePattern(ctx context.Context, pattern string) {
	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()

	iter := c.client.Scan(ctx, 0, pattern, 0).Iterator()
	deletedCount := 0
	keysToDelete := []string{}

	for iter.Next(ctx) {
		keysToDelete = append(keysToDelete, iter.Val())
	}

	if err := iter.Err(); err != nil {
		c.logger.Error("redis scan error", "error", err, "pattern", pattern)
		return
	}

	if len(keysToDelete) > 0 {
		if err := c.client.Del(ctx, keysToDelete...).Err(); err != nil {
			c.logger.Error("redis delete pattern error", "error", err, "pattern", pattern)
			return
		}
		deletedCount = len(keysToDelete)
		c.logger.Debug("deleted cache keys", "count", deletedCount, "pattern", pattern)
	}
}

func (c *RedisCache) Exists(ctx context.Context, key string) bool {
	exists, err := c.client.Exists(ctx, key).Result()
	if err != nil {
		c.logger.Error("redis exists error", "error", err, "key", key)
		return false
	}
	return exists > 0
}

func (c *RedisCache) Clear(ctx context.Context) {
	if err := c.client.FlushDB(ctx).Err(); err != nil {
		c.logger.Error("redis clear error", "error", err)
	}
}

func (c *RedisCache) Close() error {
	return c.client.Close()
}

// CreateCache creates appropriate cache backend based on configuration
func CreateCache(redisURL string, logger *slog.Logger) (Cache, error) {
	// If Redis URL is set and not default, try to use Redis
	if redisURL != "" && redisURL != "redis://localhost:6379/0" {
		redisCache, err := NewRedisCache(redisURL, logger)
		if err != nil {
			logger.Warn("Redis not available, using in-memory cache", "error", err)
			return NewInMemoryCache(), nil
		}
		return redisCache, nil
	}

	// Default to in-memory cache
	return NewInMemoryCache(), nil
}

// Init initializes the default cache instance
func Init(redisURL string, logger *slog.Logger) error {
	var err error
	once.Do(func() {
		defaultCache, err = CreateCache(redisURL, logger)
	})
	return err
}

// Get gets a value from the default cache
func Get(ctx context.Context, key string) (interface{}, bool) {
	if defaultCache == nil {
		panic("cache not initialized: call cache.Init() first")
	}
	return defaultCache.Get(ctx, key)
}

// Set sets a value in the default cache
func Set(ctx context.Context, key string, value interface{}, ttl time.Duration) {
	if defaultCache == nil {
		panic("cache not initialized: call cache.Init() first")
	}
	defaultCache.Set(ctx, key, value, ttl)
}

// Delete deletes a key from the default cache
func Delete(ctx context.Context, key string) {
	if defaultCache == nil {
		panic("cache not initialized: call cache.Init() first")
	}
	defaultCache.Delete(ctx, key)
}

// DeletePattern deletes all keys matching pattern from the default cache
func DeletePattern(ctx context.Context, pattern string) {
	if defaultCache == nil {
		panic("cache not initialized: call cache.Init() first")
	}
	defaultCache.DeletePattern(ctx, pattern)
}

// Exists checks if a key exists in the default cache
func Exists(ctx context.Context, key string) bool {
	if defaultCache == nil {
		panic("cache not initialized: call cache.Init() first")
	}
	return defaultCache.Exists(ctx, key)
}

// Clear clears all keys from the default cache
func Clear(ctx context.Context) {
	if defaultCache == nil {
		panic("cache not initialized: call cache.Init() first")
	}
	defaultCache.Clear(ctx)
}

// Close closes the default cache connection
func Close() error {
	if defaultCache == nil {
		return nil
	}
	return defaultCache.Close()
}

File: api/rust-axum/files/src/cache.rs
//! Cache Utility Module
//!
//! Provides caching functionality with Redis (optional) or in-memory fallback.
//! Install Redis support: Add `redis` crate to Cargo.toml
//!
//! Usage:
//!     use crate::cache;
//!
//!     // Set a value (TTL in seconds)
//!     cache.set("user:123", user_data, 3600).await;
//!
//!     // Get a value
//!     let user = cache.get("user:123").await;
//!
//!     // Delete a value
//!     cache.delete("user:123").await;
//!
//!     // Delete all keys matching pattern (supports * wildcard)
//!     cache.delete_pattern("user:*").await;        // All user keys
//!     cache.delete_pattern("product:*").await;     // All product keys

use moka::future::Cache as MokaCache;
use redis::AsyncCommands;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::RwLock;

/// Cache trait for cache operations
#[async_trait::async_trait]
pub trait CacheBackend: Send + Sync {
    async fn get(&self, key: &str) -> Option<String>;
    async fn set(&self, key: &str, value: &str, ttl: Duration);
    async fn delete(&self, key: &str);
    async fn delete_pattern(&self, pattern: &str);
    async fn exists(&self, key: &str) -> bool;
    async fn clear(&self);
}

/// In-memory cache implementation using Moka
pub struct InMemoryCache {
    cache: Arc<MokaCache<String, String>>,
}

impl InMemoryCache {
    pub fn new() -> Self {
        let cache = Arc::new(
            MokaCache::builder()
                .max_capacity(10_000)
                .time_to_live(Duration::from_secs(300)) // 5 minutes default TTL
                .time_to_idle(Duration::from_secs(600)) // 10 minutes idle
                .build(),
        );

        Self { cache }
    }
}

#[async_trait::async_trait]
impl CacheBackend for InMemoryCache {
    async fn get(&self, key: &str) -> Option<String> {
        self.cache.get(key).await
    }

    async fn set(&self, key: &str, value: &str, ttl: Duration) {
        let mut builder = MokaCache::builder();
        builder = builder.max_capacity(10_000);
        builder = builder.time_to_live(ttl);
        // Note: Moka doesn't support per-key TTL, so we use the default TTL
        // For exact TTL control, would need to wrap values with expiration timestamps
        self.cache.insert(key.to_string(), value.to_string()).await;
    }

    async fn delete(&self, key: &str) {
        self.cache.invalidate(key).await;
    }

    async fn delete_pattern(&self, pattern: &str) {
        // Simple wildcard support for in-memory cache
        if pattern.contains('*') {
            let parts: Vec<&str> = pattern.split('*').collect();
            let prefix = parts[0];
            let suffix = parts.get(1).unwrap_or(&"");

            // Moka doesn't have a scan iterator, so we need to track keys separately
            // For simplicity, this is a limitation - pattern matching works better with Redis
            // In practice, you'd maintain a separate index of keys or use Redis for pattern matching
            tracing::warn!("Pattern matching in in-memory cache is limited. Use Redis for full pattern support.");
        } else {
            // Exact match
            self.cache.invalidate(pattern).await;
        }
    }

    async fn exists(&self, key: &str) -> bool {
        self.cache.get(key).await.is_some()
    }

    async fn clear(&self) {
        self.cache.invalidate_all();
    }
}

/// Redis cache implementation
pub struct RedisCache {
    client: Arc<RwLock<redis::Client>>,
}

impl RedisCache {
    pub async fn new(url: &str) -> anyhow::Result<Self> {
        let client = redis::Client::open(url)?;
        let mut conn = client.get_async_connection().await?;
        redis::cmd("PING").query_async::<_, ()>(&mut conn).await?;

        Ok(Self {
            client: Arc::new(RwLock::new(client)),
        })
    }
}

#[async_trait::async_trait]
impl CacheBackend for RedisCache {
    async fn get(&self, key: &str) -> Option<String> {
        let client = self.client.read().await;
        let mut conn = match client.get_async_connection().await {
            Ok(conn) => conn,
            Err(e) => {
                tracing::error!("Redis connection error: {}", e);
                return None;
            }
        };

        match conn.get::<_, String>(key).await {
            Ok(val) => Some(val),
            Err(redis::RedisError::from((redis::ErrorKind::TypeError, "nil"))) => None,
            Err(e) => {
                tracing::error!("Redis get error: {}", e);
                None
            }
        }
    }

    async fn set(&self, key: &str, value: &str, ttl: Duration) {
        let client = self.client.read().await;
        let mut conn = match client.get_async_connection().await {
            Ok(conn) => conn,
            Err(e) => {
                tracing::error!("Redis connection error: {}", e);
                return;
            }
        };

        if let Err(e) = conn.set_ex::<_, _, ()>(key, value, ttl.as_secs() as usize).await {
            tracing::error!("Redis set error: {}", e);
        }
    }

    async fn delete(&self, key: &str) {
        let client = self.client.read().await;
        let mut conn = match client.get_async_connection().await {
            Ok(conn) => conn,
            Err(e) => {
                tracing::error!("Redis connection error: {}", e);
                return;
            }
        };

        if let Err(e) = conn.del::<_, ()>(key).await {
            tracing::error!("Redis delete error: {}", e);
        }
    }

    async fn delete_pattern(&self, pattern: &str) {
        let client = self.client.read().await;
        let mut conn = match client.get_async_connection().await {
            Ok(conn) => conn,
            Err(e) => {
                tracing::error!("Redis connection error: {}", e);
                return;
            }
        };

        let mut iter: redis::AsyncIter<String> = match conn.scan_match(pattern).await {
            Ok(iter) => iter,
            Err(e) => {
                tracing::error!("Redis scan error: {}", e);
                return;
            }
        };

        let mut keys_to_delete = Vec::new();
        while let Some(key) = iter.next_item().await {
            keys_to_delete.push(key);
        }

        if !keys_to_delete.is_empty() {
            if let Err(e) = conn.del::<_, ()>(&keys_to_delete).await {
                tracing::error!("Redis delete pattern error: {}", e);
            } else {
                tracing::debug!("Deleted {} cache keys matching pattern: {}", keys_to_delete.len(), pattern);
            }
        }
    }

    async fn exists(&self, key: &str) -> bool {
        let client = self.client.read().await;
        let mut conn = match client.get_async_connection().await {
            Ok(conn) => conn,
            Err(e) => {
                tracing::error!("Redis connection error: {}", e);
                return false;
            }
        };

        match conn.exists::<_, i32>(key).await {
            Ok(count) => count > 0,
            Err(e) => {
                tracing::error!("Redis exists error: {}", e);
                false
            }
        }
    }

    async fn clear(&self) {
        let client = self.client.read().await;
        let mut conn = match client.get_async_connection().await {
            Ok(conn) => conn,
            Err(e) => {
                tracing::error!("Redis connection error: {}", e);
                return;
            }
        };

        if let Err(e) = redis::cmd("FLUSHDB").query_async::<_, ()>(&mut conn).await {
            tracing::error!("Redis clear error: {}", e);
        }
    }
}

/// Main cache wrapper that can use either backend
pub struct Cache {
    backend: Arc<dyn CacheBackend>,
}

impl Cache {
    /// Create a cache instance (Redis if available, otherwise in-memory)
    pub async fn new(redis_url: &str) -> anyhow::Result<Self> {
        let backend: Arc<dyn CacheBackend> = if !redis_url.is_empty() && redis_url != "redis://localhost:6379/0" {
            match RedisCache::new(redis_url).await {
                Ok(redis_cache) => Arc::new(redis_cache) as Arc<dyn CacheBackend>,
                Err(e) => {
                    tracing::warn!("Redis not available, using in-memory cache: {}", e);
                    Arc::new(InMemoryCache::new()) as Arc<dyn CacheBackend>
                }
            }
        } else {
            Arc::new(InMemoryCache::new()) as Arc<dyn CacheBackend>
        };

        Ok(Self { backend })
    }

    pub async fn get(&self, key: &str) -> Option<String> {
        self.backend.get(key).await
    }

    pub async fn set(&self, key: &str, value: &str, ttl: Duration) {
        self.backend.set(key, value, ttl).await;
    }

    pub async fn delete(&self, key: &str) {
        self.backend.delete(key).await;
    }

    pub async fn delete_pattern(&self, pattern: &str) {
        self.backend.delete_pattern(pattern).await;
    }

    pub async fn exists(&self, key: &str) -> bool {
        self.backend.exists(key).await
    }

    pub async fn clear(&self) {
        self.backend.clear().await;
    }
}

// Re-export for convenience
pub use Cache as cache;

